test/core/iomgr/mpmcqueue_test.cc

```diff
@@ -0,0 +1,196 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "src/core/lib/iomgr/executor/mpmcqueue.h"
+
+#include <grpc/grpc.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/gpr/useful.h"
+#include "src/core/lib/gprpp/thd.h"
+#include "test/core/util/test_config.h"
+
+#define THREAD_SMALL_ITERATION 100
+#define THREAD_LARGE_ITERATION 10000
+
+// Testing items for queue
+struct WorkItem {
+  int index;
+  bool done;
+
+  WorkItem(int i) : index(i) { done = false; }
+};
+
+// Thread for put items into queue
+class ProducerThread {
+ public:
+  ProducerThread(grpc_core::InfLenFIFOQueue* queue, int start_index,
+                 int num_items)
+      : start_index_(start_index), num_items_(num_items), queue_(queue) {
+    items_ = nullptr;
+    thd_ = grpc_core::Thread(
+        "mpmcq_test_mt_put_thd",
+        [](void* th) { static_cast<ProducerThread*>(th)->Run(); }, this);
+  }
+  ~ProducerThread() {
+    for (int i = 0; i < num_items_; ++i) {
+      GPR_ASSERT(items_[i]->done);
+      delete items_[i];
+    }
+    delete[] items_;
+  }
+
+  void Start() { thd_.Start(); }
+  void Join() { thd_.Join(); }
+
+ private:
+  void Run() {
+    items_ = new WorkItem*[num_items_];
```

Comment:
Personally I prefer grpc_core::New and grpc_core::Delete in code under core/. This is also the reason why I ask previously whether this test needs a C++ compiler.